feat: implemented simple kafka messaging

VPesochniy opened the pull request:
Реализовал взаимодействие приложения с Kafka:

- Добавил конфигурацию топиков и producer/consumer для Kafka (KafkaTopicConfig.java, KafkaProducerConfig.java, KafkaConsumerConfig.java, KafkaPropertiesConfig.java, application.properties)
- Реализовал producer, отправляющий сообщения о новых пользователях и зарегистрированных пользователях (KafkaUserProducer.java)
- Реализовал consumer, сохраняющий пользователей в БД и отправляющий сообщения о зарегистрированных пользователях (KafkaUserConsumer.java)
- Добавил контроллер для отправки моковых пользователей и получения зарегистрированных пользователей (UserController.java)
- Добавил сервисный слой для обработки пользователей (UserService.java)
- Обновил docker-compose.yaml для запуска Kafka и Zookeeper

Commit:
tmp

Commit:
fix(UserService): fix path to MOCK_DATA.json to be able to run app as jar file
The previous implementation used an absolute path to the
MOCK_DATA.json file, which is not suitable for running the
application as a jar file.
The updated implementation uses ClassPathResource to load the
file from the classpath, which allows the application to be
run as a jar file.

Commit:
feat: add kafka support

Adds Kafka dependencies and configuration to the project.
This includes setting up a producer and consumer,
as well as defining a topic for user registration.
This allows the application to send and receive messages
from Kafka, enabling asynchronous communication and
event-driven architecture.

Commit:
feat(kafka): implement Kafka consumer

This commit introduces a Kafka consumer to the application.

The Kafka consumer listens to a topic specified in the
application.properties file and processes messages received
from that topic.

This feature enables the application to consume messages
from Kafka and react to them accordingly.

Commit:
feat(kafka): implement user registration via Kafka

This commit introduces a new feature that allows user registration through Kafka.
Instead of directly saving users to the database, the application now sends user data to a Kafka topic.
A consumer then listens to this topic, saves the users to the database, and sends a confirmation message to another Kafka topic.
This implementation improves scalability and decoupling of the application.

Additionally, the Spring Boot version was updated from 3.3.4 to 3.3.5.

Commit:
feat(kafka): add topic for registered users

Adds a new Kafka topic for registered users.
This topic will be used to publish events when a user successfully registers.
This allows other services to consume these events and perform actions based on user registration, such as sending a welcome email.

Commit:
refactor(kafka): rename KafkaProperties to KafkaPropertiesConfig

refactor(kafka): use KafkaPropertiesConfig instead of KafkaProperties

feat(kafka): add manual acknowledgment to KafkaUserConsumer

refactor(kafka): use @Payload annotation in KafkaUserConsumer

refactor(kafka): add comment to KafkaUserConsumer

refactor(kafka): add comment to User entity

refactor(kafka): add comment to UserService

Commit:
chore(docker-compose.yaml): simplify kafka and zookeeper configuration

Removes unnecessary ports, environment variables and dependencies to streamline the configuration and improve maintainability.
Simplifies Kafka advertised listeners and listener security protocol map for easier setup.
Adds hostname to postgres service for better network visibility.
Adds restart: always to zookeeper and kafka services to ensure automatic restart on failures, improving the resilience of the system.

Commit:
refactor(UserController.java): rename usersDtos variable to usersDto to better reflect that it holds a list of UserDto objects
style(UserDto.java): add a new line at the end of the file to comply with code style guidelines

Commit:
chore(docker-compose): remove KAFKA_LISTENER_SECURITY_PROTOCOL_MAP environment variable

This removes the explicit setting of the listener security protocol map, as it's not needed for a simple, single-node setup and can cause issues with newer Kafka versions.

Commit:
refactor(java): use lombok @RequiredArgsConstructor to simplify code

Use lombok @RequiredArgsConstructor annotation to automatically generate constructor for final fields.
This reduces boilerplate code and improves readability.

style(docker-compose.yaml): remove trailing whitespace

Remove trailing whitespace from docker-compose.yaml to improve code style consistency.

Alexandr056 commented on src/main/java/ru/t1/demo/kafka/amqp/kafka/KafkaUserConsumer.java:
```diff
@@ -0,0 +1,38 @@
+                .collect(Collectors.toList());
+
+        userService.saveUsersToDatabase(users);
+        ack.acknowledge();
```
а если ранее возникнет исключение? вы не отправите подтверждение вычитки сообщения, пойдете читать снова, снова упадете? (вдруг причина в наполнении сообщения)

VPesochniy replied:
Согласен, ситуация нехорошая. Лучше обернуть в блок try хотя бы и сделать кастомный exception, ну и логи прописать. Мне подумалось нелогичным подтверждать прочтение, ведь если мы не совершили полезные действия, то пусть другой consumer это сделает, раз уж мы не смогли. Если делать подтверждение прочтения, то по факту мы упали и сказали, что прочитали, т. е. просто сообщения затерли. А если отправлять снова, то кажется, что мы просто раздуваем топик.

Alexandr056 commented on src/main/java/ru/t1/demo/kafka/config/kafka/KafkaPropertiesConfig.java:
```diff
@@ -0,0 +1,25 @@
+
+@Getter
+@Configuration
+// @ConfigurationProperties /* можно использовать это, но с аннотацией @Value как-то нагляднее */
```
на счет наглядности не однозначно, но у вас структура конфига будет довольно вложенной и разрозненной, если все описывать объектами

VPesochniy replied:
Согласен, просто проект небольшой, решил так, да и с @ConfigurationProperties не работал, но обязательно учту!

Alexandr056 commented on src/main/java/ru/t1/demo/kafka/controller/UserController.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<UserDto>> getRegisteredUsers() {
```
зачем здесь ResponseEntity?
у вас всегда статус ОК, только наполняете тело, для этого есть ResponseBody или RestCotroller (который вы и используете)

Alexandr056 commented on src/main/java/ru/t1/demo/kafka/entity/User.java:
```diff
@@ -0,0 +1,46 @@
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
```
не рекомендовано над ENtity

VPesochniy replied:
Согласен, неявно может появиться сеттер для поля пароля например, но там можно указать уровень доступа. С другой стороны лучше явно указывать аннотации, либо же воспользоваться автогенерацией

Alexandr056 commented on src/main/java/ru/t1/demo/kafka/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+            parsedUsers = objectMapper.readValue(resource.getFile(), UserDto[].class);
+        } catch (IOException ex) {
+            ex.printStackTrace();
+            return Collections.emptyList();
```
не очень хорошо читается, получается что есть переменная parsedUsers, объявленная для возврата, и в зависимости от того какой ответ, так и наполняется, а тут внезапно возврат, стоит подумать как обыграть parsedUsers

Alexandr056 reviewed (approved):
хорошая работа

Commit:
refactor(java): improve code style and add exception handling

The changes in this commit improve the code style and add exception handling to the Kafka consumer.

- Changed the case of the port variable in server.ts to uppercase to improve semantics.
- Added support for the process.env.PORT environment variable to be able to run the app on a configurable port.
- Added exception handling to the Kafka consumer to prevent the application from crashing if an error occurs.
- Improved the code style of the Kafka consumer by using the Lombok library.
- Updated the dependencies in the pom.xml file.

Commit:
refactor(kafka): improve code style and optimize imports

The changes in this commit are focused on improving the code style and optimizing imports across multiple files in the Kafka module.

Specifically, the following improvements were made:

- Removed unnecessary imports and organized existing ones to enhance readability.
- Replaced `.stream().collect(Collectors.toList())` with `.toList()` for conciseness and improved performance, leveraging Java 16+ features.
- Removed redundant comments that didn't add significant value to the code understanding.
- Improved code formatting and organization for better clarity.
- Updated the `User` entity to explicitly define `hashCode` and `equals` methods, ensuring proper object comparison and behavior in collections.
- Removed the `@EqualsAndHashCode` annotation from the `User` entity as it's now handled manually.
- Added explicit type parameters to the `ConcurrentKafkaListenerContainerFactory` bean for better type safety.
- Removed an unused comment related to `@ConfigurationProperties` in `KafkaPropertiesConfig`.

These changes aim to enhance the overall quality and maintainability of the codebase by adhering to best practices and improving code clarity.
